### PR TITLE
FEATURE: Allow changing slug on create channel

### PR DESCRIPTION
**Pull request by martin-brennan**
This commit allows us to set the channel slug when creating new chat
channels. As well as this, it introduces a new `SlugsController` which can
generate a slug using `Slug.for` and a name string for input. We call this
after the user finishes typing the channel name (debounced) and fill in
the autogenerated slug in the background, and update the slug input
placeholder.

This autogenerated slug is used by default, but if the user writes anything
else in the input it will be used instead.

### CODE REVIEW

File: plugins/chat/spec/system/create_channel_spec.rb [hunk 1]
```diff
@@ -72,45 +78,75 @@
...
 
+    it "autogenerates slug from name and changes slug placeholder" do
+      chat_page.visit_browse
+      chat_page.new_channel_button.click
+      name = "Cats & Dogs"
+      channel_modal.select_category(category_1)
+      channel_modal.fill_name(name)
+      channel_modal.fill_description("All kind of cute cats")
+
+      wait_for_attribute(channel_modal.slug_input, :placeholder, "cats-dogs")
```
Comment (martin-brennan) [inline plugins/chat/spec/system/create_channel_spec.rb:99]: Looking for an opinion on this, not sure if it's crazy to do but capybara does not have an inbuilt way to wait for an attribute value.

File: app/controllers/slugs_controller.rb [hunk 2]
```diff
@@ -0,0 +1,22 @@
+# frozen_string_literal: true
+
+class SlugsController < ApplicationController
```
Comment (lis2) [inline app/controllers/slugs_controller.rb:3]: Do we even need that endpoint? One benefit is that we are using Slug for Topic/Categories/Badges so consistency but on the other hand, if we just dasherize name to slug with JS - the code would be faster and simpler. What is the biggest benefit of using Slug.for?

**Comment by lis2** [comment]
Another question, do we need some validation if slug is not already taken?

File: app/controllers/slugs_controller.rb (hunk 2 above)
Comment (martin-brennan) [reply app/controllers/slugs_controller.rb:3]: The main benefit is consistency between server and client, so using the same name produces the same result whether we are generating the slug on the server or client. `Slug.for` does a lot, and having to recreate this in JS will just introduce code drift. Eventually we should move those other places in the UI to do the same.

**Comment by martin-brennan** [comment]
> Another question, do we need some validation if slug is not already taken?

@lis2 we already do this in the `ChatChannel` validation code, however I will add a system spec in this PR to make sure the modal shows a nice error message if this occurs. I think it's _generally_ unlikely that people will come up with the same slug as existing channels often, so I would rather keep that additional DB check out of the slug endpoint for now, since it will also introduce complexity (having to know which "type" of model we are generating the slug for and checking uniqueness)

File: app/controllers/slugs_controller.rb (hunk 2 above)
Comment (lis2) [reply app/controllers/slugs_controller.rb:3]: Thank you :)

**Review by lis2 — APPROVED** [review]

File: app/controllers/slugs_controller.rb [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+  requires_login
+
+  MAX_SLUG_GENERATIONS_PER_MINUTE = 20
+
+  def generate
```
Comment (jjaffeux) [inline app/controllers/slugs_controller.rb:8]: I think we could have gone with `create` here and a POST route. I understand it's not exactly creating an object on the server, but in a way we are even though it's not persisted and it's just generally more expected (at least from my point of view).

File: app/controllers/slugs_controller.rb [hunk 4]
```diff
@@ -0,0 +1,22 @@
+# frozen_string_literal: true
+
+class SlugsController < ApplicationController
+  requires_login
+
+  MAX_SLUG_GENERATIONS_PER_MINUTE = 20
```
[the remark this answers is not in the document]
Comment (jjaffeux) [reply app/controllers/slugs_controller.rb:6]: Looks enough to me

**Pull request merged**